pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go

```diff
@@ -658,7 +683,32 @@ func addColumnsForSecondaryIndex(
 			}
 		}
 		idxSpec.columns = append([]*scpb.IndexColumn{indexColumn}, idxSpec.columns...)
-		panic(scerrors.NotImplementedErrorf(n.Sharded, "split and scatter support are missing"))
+		// Create a new check constraint for the hash sharded index
+		if shardColumnCreated {
+			checkConstraintBucketValues := strings.Builder{}
+			checkConstraintBucketValues.WriteString(fmt.Sprintf("%q IN (", shardColName))
+			for bucket := int32(0); bucket < buckets; bucket++ {
+				checkConstraintBucketValues.WriteString(strconv.Itoa(int(bucket)))
+				if bucket != buckets-1 {
+					checkConstraintBucketValues.WriteString(",")
+				}
+			}
+			checkConstraintBucketValues.WriteString(")")
+			_, err = parser.ParseExpr(checkConstraintBucketValues.String())
+			if err != nil {
+				panic(err)
+			}
+			shardCheckConstraint := &scpb.CheckConstraint{
+				TableID:      idxSpec.secondary.TableID,
+				ConstraintID: idxSpec.secondary.ConstraintID + 2,
```

Comment:
Can we avoid relying on magical offsets here? Also why isn't this using `b.NextTableConstraintID`?